### PR TITLE
feat: generic contract calls in Account contract

### PR DESCRIPTION
**Pull request by gabspeck**
* Reorganize Giano packages
* Fix contract tests
* Create JS client for Giano account contracts
* Change `execute` signature payload to expect encoded call data
* Fix TypeScript type errors

Copilot-generated summary follows below

---

This pull request includes significant changes to the `giano` project, including a switch from Yarn to pnpm, new client package functionalities, and substantial updates to the common package. Below are the most important changes grouped by theme:

### Package Management Updates:
* [`.yarnrc.yml`](diffhunk://#diff-88fbe28c4102501b94961511a0d70ff895bf39970b4d3fc11917794a239117c5L1): Removed the `nodeLinker` configuration.
* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L3-R3): Switched package manager from `yarn@4.1.0` to `pnpm@8.15.6`.

### Client Package Enhancements:
* [`packages/client/package.json`](diffhunk://#diff-26d3d28d31824ef26252df77cca08d24faea8451cb8fd3ffee2000f9e496daa0R1-R13): Added a new `package.json` for the client package with dependencies and metadata.
* [`packages/client/index.ts`](diffhunk://#diff-a46a255a83eb275f01a88e86319b2f5ba010d530a5cbbdab9b1122dff03a35ebR1): Exported all modules from `src/wallet`.
* [`packages/client/src/wallet/index.ts`](diffhunk://#diff-9a311b802e67cd4bc4f5f7133ade5b8e3afe2c054e712a2b9d6f51c981dfcce4R1): Exported all modules from `walletClient`.
* [`packages/client/src/wallet/walletClient.ts`](diffhunk://#diff-0580301efb3d94214b4a29ccf2084a0eeb38e7bb34699bea9b1bd9241e947639R1-R83): Implemented `GianoWalletClient` with methods for account management and transaction proxying.

### Common Package Refactor:
* [`packages/common/package.json`](diffhunk://#diff-50d7c39a9430d37971aa76858165ab4f7921c4cc4340b28e9b673ce6982e63cfL3-R11): Added a new `package.json` for the common package with dependencies and metadata.
* [`packages/common/src/encoding/encodeChallenge.ts`](diffhunk://#diff-81bfd61f2dc4980b5545254133a271cbba59045262035ab4247c7ff2ae14f000R1-R25): Added `encodeChallenge` function to encode challenge data.
* [`packages/common/src/encoding/index.ts`](diffhunk://#diff-1813e1d38149520e3170a957109a7c31aed6c98fa1cd4138609e71a97c58b22cR1-R2): Exported `encodeChallenge` and `numbers` modules.
* [`packages/common/src/encoding/numbers.ts`](diffhunk://#diff-5e6168e2e5aabb5717fe82ab51c499e07c13e30249a0c5a7086a8df6fab66fd1L1-R13): Refactored utility functions for number encoding.
* Removed multiple files and functions related to ASN.1 parsing, signature verification, and helpers. [[1]](diffhunk://#diff-608baccc6a63781bc3f41775544bb9a423ab27d080d4639283c0be6bde725a24L1-L41) [[2]](diffhunk://#diff-aab572f1e19def020d112fbfb0832569e86cad192d857817c3a27fc6a89c2990L1-L10) [[3]](diffhunk://#diff-7dbd97dd834ddd037fa619e2aaf4a8d1034a14a6c425ed4bab39663799f0bf88L1-L13) [[4]](diffhunk://#diff-985932c9fdc65530d3635369b116da51fcc7c1e16bdef7cfa3105fc830cd727bL1-L105) [[5]](diffhunk://#diff-09a3c654592ee1ebc9be47e1038541c0a65cc0e9935390fa65c563552b01776bL1-L82) [[6]](diffhunk://#diff-6b82a4a1c3c80a091d0756c9a9217b7cd619965e0d99237b56650f5684af0dbfL1-L43)

### Contract Updates:
* [`packages/contracts/contracts/Account.sol`](diffhunk://#diff-1f8d126340697bb5a0076e3651c7a25571ade3f9460435108192af5748da082dL4-R5): Updated the `Account` contract to include `ReentrancyGuard`, added `Call` struct, and modified the `getChallenge` function to include call data. [[1]](diffhunk://#diff-1f8d126340697bb5a0076e3651c7a25571ade3f9460435108192af5748da082dL4-R5) [[2]](diffhunk://#diff-1f8d126340697bb5a0076e3651c7a25571ade3f9460435108192af5748da082dR16-R26) [[3]](diffhunk://#diff-1f8d126340697bb5a0076e3651c7a25571ade3f9460435108192af5748da082dL36-R43) [[4]](diffhunk://#diff-1f8d126340697bb5a0076e3651c7a25571ade3f9460435108192af5748da082dL50-R56)

### CODE REVIEW

**Review by kozaqui — COMMENTED** [review]
Looking really good 💪 
However I'm worried about `getChallenge` and `validSignature` having a potential vulnerability, where the Passkey user signature is not required for deciding the `target` contract and the `value` of ETH that it would be sending to it.

**Comment by kozaqui** [comment]
+ another thing. I'm wondering if it would be beneficial to implement `IERC721Receiver` interface, to allow `safeTransferFrom` to send ERC721 tokens to the Giano Account as well (and not only `transferFrom`)

File: packages/contracts/contracts/Account.sol [hunk 1]
```diff
@@ -64,14 +67,19 @@ contract Account {
...
     receive() external payable {}
-
     // solhint-disable-next-line no-empty-blocks
     fallback() external payable {}
 
-    function _validateSignature(
-        bytes memory message,
-        bytes calldata signature
-    ) private view returns (bool) {
+    function execute(Call calldata call) external payable validSignature(bytes.concat(getChallenge(call.data)), call.signature) nonReentrant {
```
[the remark this answers is not in the document]
Comment (gabspeck) [reply packages/contracts/contracts/Account.sol:73]: Very good point, we should hash all parameters definitely. Also, I noticed something even worse: the nonce is not being incremented anywhere! Time to fix that 😅

**Review by kozaqui — APPROVED** [review]
Right, the `currentNonce++` was an important bit as well 😅 

Also - not necessarily for this PR, but can you consider: https://github.com/appliedblockchain/giano/pull/8#issuecomment-2444664022 ?

**Comment by gabspeck** [comment]
> Right, the `currentNonce++` was an important bit as well 😅
> 
> Also - not necessarily for this PR, but can you consider: [#8 (comment)](https://github.com/appliedblockchain/giano/pull/8#issuecomment-2444664022) ?

I was taking a look at this and I concluded that it's probably best to keep this contract as basic as possible and allow it to be extended with extra features if needed. So if my project needs ERC721 or ERC1155 safeTransferFrom support for instance, I'd create a new one that extends Giano's `Account`. What do you think?

**Comment by kozaqui** [comment]
I like this approach. How about we do create this extended contract in Giano repo as well - that implements some of the common interfaces to be used in prospective projects?

**Review by stvkoch — APPROVED** [review]
LGTM

**Pull request merged**